### PR TITLE
Add mani cli tool

### PR DESCRIPTION
**Pull request by thornycrackers**
###### Motivation for this change

Add the mani cli tool

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [*] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I don't really know much GO so I mostly copied and pasted. The tests were trying to create files and got permission denied so I just skipped them for now. The resulting binary seems to be working fine, haven't had any issues yet.

### CODE REVIEW

**Comment by teto** [comment]
I've been wanting such a tool for some time (I think it fits the bill). Would be nice to have autocompletion scripts installed too. (grep nixpkgs for installShellFile)

File: pkgs/development/tools/mani/default.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    owner = "alajmo";
+    repo = "mani";
+    rev = "v${version}";
+    sha256 = "sha256-9rcgPeYFHdIN73K0zGPEHqFFLFkVYkNYRXJ+0/Zo4zI=";
+  };
+
+  vendorSha256 = "sha256-ZivzDfjx2djzS0Xm3GISK3zpB5fUUMgy2o4Ti1Z9wMM=";
+
+  # Skip tests
+  checkPhase = null;
```
Comment (legendofmiracles) [inline pkgs/development/tools/mani/default.nix:17]:
```suggestion
  # Skip tests
  doCheck = false;
```
Can you please add a comment why we are skipping the checks?

File: pkgs/development/tools/mani/default.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    longDescription = ''
+      mani is a CLI tool that helps you manage multiple repositories. It's useful
+      when you are working with microservices, multi-project systems, many
+      libraries or just a bunch of repositories and want a central place for
+      pulling all repositories and running commands over them.
+    '';
+    homepage = "https://manicli.com/";
+    changelog = "https://github.com/alajmo/mani/releases/tag/v${version}";
+    license = licenses.mit;
+    platforms = platforms.unix;
```
Comment (legendofmiracles) [inline pkgs/development/tools/mani/default.nix:30]:
```suggestion
```
Set by buildGoModule

**Comment by legendofmiracles** [comment]
This should also depend on git, no? You should add it as a `buildInput` and then wrap mani to also find it at runtime

File: pkgs/development/tools/mani/default.nix (hunk 1 above)
Comment (thornycrackers) [reply pkgs/development/tools/mani/default.nix]: Updated

File: pkgs/development/tools/mani/default.nix (hunk 2 above)
Comment (thornycrackers) [reply pkgs/development/tools/mani/default.nix]: updated

File: pkgs/development/tools/mani/default.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  buildInputs = [ git ];
+
+  postInstall = ''
+    # generate completion scripts for mani
+    export HOME="$NIX_BUILD_TOP"
+    mkdir -p $out/share/{bash-completion/completions,fish/vendor_completions.d,zsh/site-functions}
+    $out/bin/mani completion bash > "$out/share/bash-completion/completions/mani"
+    $out/bin/mani completion fish > $out/share/fish/vendor_completions.d/mani.fish
+    $out/bin/mani completion zsh > $out/share/zsh/site-functions/_mani
```
Comment (teto) [inline pkgs/development/tools/mani/default.nix:24]: Though that works, the standard  nixpkgs way uses `installShellCompletion`.

**Comment by thornycrackers** [comment]
@teto That's a great suggestion. I added a snippet that I found in another go package called `jump` to generate the autocomplete.

@legendofmiracles  Updates applied from your suggestions.

Thank you both for the review. I am struggling to verify that the changes work correctly though. For the autocomplete I tried `nix-env -f $NIXPKGS -iA mani` but I wasn't able to get the autocomplete to work. To test the git change I tried uninstalling git locally and then adding the `buildInputs` but mani wouldn't actually clone repos unless I specifically had git installed. I'm still fairly new to packaging so your feedback is appreciated.

**Comment by legendofmiracles** [comment]
> completion doesn’t work

That’s because your bash is configured to not load completion scripts. It works here where I enabled it in fish.

> git needs to be installed

This patch should fix it: http://0x0.st/-RV-.txt

Another note, your commits and PR titles should follow the contributing guidelines, so: “mani: init at <version>”
and squash the other package related commit

File: pkgs/development/tools/mani/default.nix (hunk 3 above)
Comment (thornycrackers) [reply pkgs/development/tools/mani/default.nix]: I figured out why I couldn't get it working before. Updated.

**Comment by thornycrackers** [comment]
Updated with the diff and tested mani and it does work now with git not installed, thankyou. I thought that's what `buildInputs` was for? how do I know in the future when to use buildInputs and when to wrap with wrapProgram?

Commits squashed and message updated.

Also updated the completions to use `installShellCompletion` instead of what I copied previously

again, thank you for the review.

File: pkgs/development/tools/mani/default.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  vendorSha256 = "sha256-ZivzDfjx2djzS0Xm3GISK3zpB5fUUMgy2o4Ti1Z9wMM=";
+
+  nativeBuildInputs = [ installShellFiles makeWrapper ];
+
+  postInstall = ''
+    installShellCompletion --cmd dyff \
+      --bash <($out/bin/dyff completion bash) \
+      --fish <($out/bin/dyff completion fish) \
+      --zsh <($out/bin/dyff completion zsh)
+    wrapProgram $out/bin/mani \
```
Comment (legendofmiracles) [inline pkgs/development/tools/mani/default.nix:23]:
```suggestion
    
    wrapProgram $out/bin/mani \
```
Minor nitpick

**Review by legendofmiracles — APPROVED** [review]
> I thought that's what buildInputs was for? how do I know in the future when to use buildInputs and when to wrap with wrapProgram?

You can actually not specify runtime dependencies via nixpkgs (at least not without makeWrapper). What you are specifying are libraries to be linked, and it only differs from nativeBuildInputs (which is often conceived as build dependencies) in a cross compiling sense.

File: pkgs/development/tools/mani/default.nix (hunk 4 above)
Comment (thornycrackers) [reply pkgs/development/tools/mani/default.nix:24]: updated.

**Review by legendofmiracles — APPROVED** [review]

File: pkgs/development/tools/mani/default.nix [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      --fish <($out/bin/mani completion fish) \
+      --zsh <($out/bin/mani completion zsh)
+
+    wrapProgram $out/bin/mani \
+      --prefix PATH : ${lib.makeBinPath [ git ]}
+  '';
+
+  # Skip tests
+  # The repo's test folder has a README.md with detailed information. I don't
+  # know how to wrap the dependancies for these integration tests so skip for now.
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mani/default.nix:30]:
```suggestion
  # know how to wrap the dependencies for these integration tests so skip for now.
```

Comment (thornycrackers) [reply pkgs/development/tools/mani/default.nix]: updated.

**Pull request merged**